File: service/src/main/java/coding/challenge/core/i18n/Messages.java
package coding.challenge.core.i18n;

import coding.challenge.core.context.ExecutionContext;
import lombok.experimental.UtilityClass;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.StringUtils;

import java.util.HashMap;
import java.util.Locale;
import java.util.Map;
import java.util.ResourceBundle;
import java.util.Set;

/**
 * Contains a set of constants with identifiers of resource messages. Provides method to get a message from resource bundle.
 */
@UtilityClass
@Slf4j
public class Messages {

    public final String FAILED_TO_RETRIEVE_FILE = "failed_to_retrieve_file";

    public final String FAILED_TO_SAVE_FILE = "failed_to_save_file";

    public final String FILE_NOT_FOUND = "file_not_found";

    public final String INTERNAL_ERROR = "internal_error";

    public final String INVALID_ARGUMENT = "invalid_argument";

    public final String PHRASE_NOT_FOUND = "phrase_not_found";

    public final String UNSUPPORTED_AUDIO_TYPE = "unsupported_audio_type";

    public final String USER_NOT_FOUND = "user_not_found";

    private final Set<Locale> SUPPORTED_LOCALES = Set.of(Locale.ENGLISH);

    private final Map<Locale, ResourceBundle> RESOURCE_BUNDLES = new HashMap<>();

    private final ResourceBundle DEFAULT_RESOURCE_BUNDLE;

    static {
        for (Locale locale : SUPPORTED_LOCALES) {
            RESOURCE_BUNDLES.put(locale, ResourceBundle.getBundle("messages", locale));
        }
        DEFAULT_RESOURCE_BUNDLE = ResourceBundle.getBundle("messages");
    }

    /**
     * Returns text of the message that is mapped to the key based on the locale. Can accept parameters to format the text using {@link String#format}
     *
     * @param key        The key of the message
     * @param parameters Parameters for the message
     * @return Text from resource bundle or empty string
     */
    public String getText(String key, Object... parameters) {
        Locale locale = Locale.forLanguageTag(ExecutionContext.get().getLanguage());
        ResourceBundle resourceBundle = RESOURCE_BUNDLES.get(locale);
        if (resourceBundle == null) {
            log.warn("Resource bundle for locale {} was not found", locale);
            return StringUtils.EMPTY;
        }
        String message;
        if (resourceBundle.containsKey(key)) {
            message = resourceBundle.getString(key);
        } else if (DEFAULT_RESOURCE_BUNDLE.containsKey(key)) {
            message = DEFAULT_RESOURCE_BUNDLE.getString(key);
        } else {
            log.warn("Resource string for key {} was not found", key);
            return StringUtils.EMPTY;
        }
        return String.format(message, parameters);
    }

}

File: service-kt/src/main/kotlin/coding/challenge/core/i18n/Messages.kt
package coding.challenge.core.i18n

import coding.challenge.core.context.ExecutionContext
import org.apache.commons.lang3.StringUtils
import org.slf4j.LoggerFactory
import java.util.Locale
import java.util.ResourceBundle

/**
 * Contains a set of constants with identifiers of resource messages. Provides method to get a message from resource bundle.
 */
object Messages {

    const val FAILED_TO_RETRIEVE_FILE: String = "failed_to_retrieve_file"

    const val FAILED_TO_SAVE_FILE: String = "failed_to_save_file"

    const val FILE_NOT_FOUND: String = "file_not_found"

    const val INTERNAL_ERROR: String = "internal_error"

    const val INVALID_ARGUMENT: String = "invalid_argument"

    const val PHRASE_NOT_FOUND: String = "phrase_not_found"

    const val UNSUPPORTED_AUDIO_TYPE: String = "unsupported_audio_type"

    const val USER_NOT_FOUND: String = "user_not_found"

    private val log = LoggerFactory.getLogger(this::class.java)

    private val SUPPORTED_LOCALES: Set<Locale> = setOf(Locale.ENGLISH)

    private val RESOURCE_BUNDLES: MutableMap<Locale, ResourceBundle> = mutableMapOf()

    private val DEFAULT_RESOURCE_BUNDLE: ResourceBundle

    init {
        for (locale in SUPPORTED_LOCALES) {
            RESOURCE_BUNDLES[locale] = ResourceBundle.getBundle("messages", locale)
        }
        DEFAULT_RESOURCE_BUNDLE = ResourceBundle.getBundle("messages")
    }

    /**
     * Returns text of the message that is mapped to the key based on the locale. Can accept parameters to format the text using [String.format]
     *
     * @param key        The key of the message
     * @param parameters Parameters for the message
     * @return Text from resource bundle or empty string
     */
    fun getText(key: String, vararg parameters: Any?): String {
        val locale = Locale.forLanguageTag(ExecutionContext.get().language)
        val resourceBundle = RESOURCE_BUNDLES[locale]
        if (resourceBundle == null) {
            log.warn("Resource bundle for locale {} was not found", locale)
            return StringUtils.EMPTY
        }
        val message =
            if (resourceBundle.containsKey(key)) {
                resourceBundle.getString(key)
            } else if (DEFAULT_RESOURCE_BUNDLE.containsKey(key)) {
                DEFAULT_RESOURCE_BUNDLE.getString(key)
            } else {
                log.warn("Resource string for key {} was not found", key)
                return StringUtils.EMPTY
            }
        return String.format(message, *parameters)
    }

}
